src/promptflow/promptflow/executor/_line_execution_process_pool.py

```diff
@@ -207,65 +180,121 @@ def __exit__(self, exc_type, exc_val, exc_tb):
+                if not completed:
+                    self.handle_line_timeout(line_number, timeout_time, inputs, run_id, start_time, result_list)
+                elif crashed:
+                    self.handle_process_crashed(line_number, inputs, run_id, start_time, result_list)
+
+                self._completed_idx[line_number] = format_current_process(
+                    process_name, pid, line_number, is_failed=True)
                 if not task_queue.empty():
-                    healthy_ensured_process.end()
-                    healthy_ensured_process.start_new(task_queue)
+                    self._processes_manager.restart_process(index)
+                    process_info = self._get_process_info(index, input_queue, output_queue)
+            else:
+                # Handle line execution completed.
+                self._completed_idx[line_number] = format_current_process(
+                    process_name, pid, line_number, is_completed=True)
 
             self._processing_idx.pop(line_number)
-            log_progress(
-                run_start_time=run_start_time,
-                logger=bulk_logger,
-                count=len(result_list),
-                total_count=self._nlines,
-            )
+
+    def _get_process_info(self, index, input_queue, output_queue):
+        while True:
+            try:
+                pid = self._process_info[index]['pid']
+                process_name = self._process_info[index]['process_name']
+                process_info = (index, pid, process_name, input_queue, output_queue)
+                return process_info
+            except KeyError:
+                continue
+
+    def _process_message(self, message, result_list):
+        if isinstance(message, LineResult):
+            message = self._process_multimedia(message)
+            result_list.append(message)
+            return True
+        elif isinstance(message, FlowRunInfo):
+            self._storage.persist_flow_run(message)
+        elif isinstance(message, NodeRunInfo):
+            self._storage.persist_node_run(message)
+
+        return False
+
+    def handle_line_timeout(self, line_number, timeout_time, inputs, run_id, start_time, result_list):
+        bulk_logger.warning(f"Line {line_number} timeout after {timeout_time} seconds.")
+        ex = LineExecutionTimeoutError(line_number, timeout_time)
+        result = self._generate_line_result_for_exception(
+            inputs, run_id, line_number, self._flow_id, start_time, ex
+        )
+        result_list.append(result)
+
+    def handle_process_crashed(self, line_number, inputs, run_id, start_time, result_list):
+        bulk_logger.warning(f"Process crashed while executing line {line_number},")
+        ex = ProcessCrashError(line_number)
+        result = self._generate_line_result_for_exception(
+            inputs, run_id, line_number, self._flow_id, start_time, ex
+        )
+        result_list.append(result)
+
+    def _timeout_process_wrapper(
```

Comment:
We may have a more meaningful name of this function and better comments.

Reply:
Renamed